File: language-testing/guess_number.hs
module Main where

import qualified System.Random as R
import qualified System.IO as IO

takeGuess :: Int -> Int -> IO ()
takeGuess solution num = do
  putStr $ "Guess " ++ show num ++ ": "
  IO.hFlush IO.stdout
  guess <- readLn :: IO Int
  guessNumber guess solution num

guessNumber :: Int -> Int -> Int -> IO ()
guessNumber guess solution num
  | guess == solution = putStrLn "Correct! You have won!"
  | num == 5 = putStrLn $ "Haha, I won! The number was " ++ show solution
  | guess > solution = do 
    putStrLn "Too high! Try again!"
    takeGuess solution $ succ num
  | otherwise = do
    putStrLn "Too low! Try again!"
    takeGuess solution $ succ num

main :: IO ()
main = do
  solution <- R.randomRIO(1, 100) :: IO Int
  putStrLn "Guess-a-number game!"
  putStrLn "I am thinking of a number between 1 and 100"
  putStrLn "You have 5 tries to guess it correctly or I win"
  putStrLn "What's your guess?"
  takeGuess solution 1
  
-- TAIL INFO:
-- Name: Guess number
-- Language: Haskell
-- Compile: ghc --make guess_number.hs
-- State: Done
--
-- Play guess-a-number game
--
-- Example: ./guess_number

File: language-testing/guess_number.rs
// Written in rust 0.9
// Jan 21 2014

use std::rand::random;
use std::io::stdio::{stdin, flush};
use std::io::buffered::BufferedReader;

fn get_rand() -> int {
  let solution = random::<int>() % 100 + 1;
  if solution < 0 { -solution }
  else            {  solution }
}

fn main() {
  let mut reader = BufferedReader::new(stdin());
  let solution = get_rand();
  let mut count = 0;

  println("Guess-a-number game!\n\
  I am thinking of a number between 1 and 100\n\
  You have 5 tries to guess it correctly or I win\n\
  What's your guess?");

  loop {
    count += 1;
    print!("Guess {}: ", count);
    flush();
    let strline = reader.read_line().unwrap_or(~"");
    match from_str::<int>(strline.trim()) {
      Some(num) => {
        if num == solution {
          println("Correct! You won!");
          return;
        } else if count == 5 {
          println!("Haha, I won! The number was {}", solution);
          return;
        } else if num < solution {
          println("Too low! Try again!");
        } else {
          println("Too high! Try again!");
        }
      },
      None => println("Hey, that's not a number!")
    }
  }
}
